### PR TITLE
lncli+lnrpc: add hints flag to addinvoice command

### LINKED ISSUE #7305: [bug]: Not all hop hints are included in private invoices
### Background

Ever since #6236, the route hint selection algorithm is sub-optimal for situations where you only have a couple of private channels. 

For example, users of Blixt Wallet will usually only have around 1-5 channels.
When the number of channels is that small, I would like to argue that it's better to include all channels rather a couple of them.
From our testing, you would usually only get a single hop hint included out of many, with the current hop hint selection algorithm (hop hint for every channel would be included with the old algorithm).

Now, I understand the rationale behind the change and it makes sense to not all hop hints when you have tens och hundreds of private channels.
But if you have just a couple of ones, I think it's reasonable to include all of them because of

* Reliability. Making sure to include hop hints to different channel peers increases the chance of the payment going through
* Privacy is somewhat solved by the alias-scid feature (and potentially route blinding in the future). Thus I think prioritizing reliability over privacy is sensible.
* Only including 1 hop hint goes against MPP strategies. A payer may would like to try splitting the payment to different shards to increase payment success

### Steps to reproduce

1. Open a couple of private channels.
2. Create a private invoice with some amount (1 sat would do)
3. Decode and the invoice and observe that only 1 hop hint would be included

### Expected behaviour

I suggest that 6 hop hints should always be included when creating a private invoice.
Around 5-6 is the upper limit until QR codes become unscannable. So mobile wallets users should stay below that number (which is reasonable, no need for more than that).

We may want to make this more flexible and less opinionated than the proposed fix. But this can potentially do as a quick fix.

### Actual behaviour

Usually you will only get 1 hop hint because you would easily satisfy the hop hint selection rules:

```
we will add up to 20 channels, or enough hints to each incoming bandwidth 2 * invoice amount
```

### PR DESCRIPTION
**Pull request by hotorb-code**
## Change Description
This PR introduces the new flag `hints` for the `addinvoice` command. `hints` sets the number of routing hints the invoice should include. Poor channel liquidity may result in more hints than specified being added, lack of private channels in less. The maximum number of hints stays at 20. The `private` flag behavior is unchanged and keeps its focus on privacy. The `private`, `hints`, `amt` and `amt_msat` flags can be used together or separately depending on the desired result.

Fixes #7305.

## Steps to Test
Create a private invoice with enough private channels.

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [x] Any new logging statements use an appropriate subsystem and logging level.
- [x]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

**Comment by positiveblue** [comment]
Thank you for taking care of this @hotorb-code  :tada:

Instead of adding a magic numbe (`15`) that comes from two other implicit ones (we want at least 5 hints if we have enough peers and we currently set a maximum of 20) we should add a new constant `minNumOfHopHints` + a new param to `sufficientHints(currentNumberOfHints)`. Then inside the function we can simply compare if we did not added enough hints yet before continuing with the rest of the checks.


We will also have to cover this logic in [the tests](https://github.com/lightningnetwork/lnd/blob/fdb94afecbc01a37f1d184277bcaa18929f23a50/lnrpc/invoicesrpc/addinvoice_test.go#L505)

**Review by positiveblue — CHANGES_REQUESTED** [review]
Thank you for taking care of this @hotorb-code  :tada:

Instead of adding a magic numbe (`15`) that comes from two other implicit ones (we want at least 5 hints if we have enough peers and we currently set a maximum of 20) we should add a new constant `minNumOfHopHints` + a new param to `sufficientHints(currentNumberOfHints)`. Then inside the function we can simply compare if we did not added enough hints yet before continuing with the rest of the checks.


We will also have to cover this logic in [the tests](https://github.com/lightningnetwork/lnd/blob/fdb94afecbc01a37f1d184277bcaa18929f23a50/lnrpc/invoicesrpc/addinvoice_test.go#L505)

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thanks for the PR! However I don't think this is the right approach. We should let the user to decide the tradeoff between reliability(more hints) and privacy(fewer hints).

File: lnrpc/invoicesrpc/addinvoice.go [hunk 1]
```diff
@@ -38,6 +38,10 @@ const (
 	// inbound capacity we want our hop hints to represent, allowing us to
 	// have some leeway if peers go offline.
 	hopHintFactor = 2
+	
+	// minHopHints is the minimum number of hint paths that we try to include
+	// in an invoice if possible.	
+	minHopHints = 5
```
[the remark this answers is not in the document]
Comment (hotorb-code) [reply lnrpc/invoicesrpc/addinvoice.go:44]: If there are fewer than 5 channels then only those channels will have hop hints in the invoice. The `minHopHints` constant is only used in the `sufficientHints` function to make sure it at least tried to add 5 hop hints. I will make the comment clearer.

**Comment by hotorb-code** [comment]
Thanks for the reviews. This is turning into a different PR so I have edited the title and description. The new approach is to add a `hints` flag to the `addinvoice` command.

File: lnrpc/lightning.proto [hunk 2]
```diff
@@ -3541,6 +3541,13 @@ message Invoice {
     Note: Output only, don't specify for creating an invoice.
     */
     map<string, AMPInvoiceState> amp_invoice_state = 28;
+	
+    /*
+    The number of routing hints the invoice should include. Poor channel
+	liquidity may result in more hints being added, lack of private channels
+	in less. The maximum is 20 hints.
+    */
+    int64 hints_target = 29;
```
Comment (yyforyongyu) [inline lnrpc/lightning.proto:3550]: hmm this commit doesn't look right...did you run `make rpc-check`?

File: lnrpc/invoicesrpc/addinvoice.go [hunk 3]
```diff
@@ -136,6 +136,9 @@ type AddInvoiceData struct {
 	// RouteHints are optional route hints that can each be individually
 	// used to assist in reaching the invoice's destination.
 	RouteHints [][]zpay32.HopHint
+	
+	// How many route hints the invoice should include. Default is 0.
```
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice.go:140]:
```suggestion
	// The number of route hints the invoice should include. Default is 0.
```

File: cmd/lncli/cmd_invoice.go [hunk 4]
```diff
@@ -67,7 +67,15 @@ var addInvoiceCommand = cli.Command{
 				"payer in reaching you. If amt and amt_msat " +
 				"are zero, a large number of hints with " +
 				"these channels can be included, which " +
-				"might not be desirable.",
+				"might not be desirable. This can be limited" +
+				"with the hints flag.",
+		},
+		cli.Int64Flag{
+			Name: "hints",
```
Comment (yyforyongyu) [inline cmd/lncli/cmd_invoice.go:74]: Generally we just give the same name for a param that's used in cli/rpc.

File: docs/release-notes/release-notes-0.16.1.md [hunk 5]
```diff
@@ -60,7 +64,7 @@ https://github.com/lightningnetwork/lnd/pull/7359)
 * A [debug log](https://github.com/lightningnetwork/lnd/pull/7514) has been
   added to `lnrpc` so the node operator can know whether a certain request has
   happened or not.
-* [Add peer_scid_alias to the response of 
+* [Add peer_scid_alias to the response of
```
Comment (yyforyongyu) [inline docs/release-notes/release-notes-0.16.1.md:67]: nit: maybe don't touch unrelated code.

File: lnrpc/invoicesrpc/addinvoice.go [hunk 6]
```diff
@@ -136,6 +136,9 @@ type AddInvoiceData struct {
...
+	HintsTarget int64
```
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice.go:141]: how about `NumRouteHints`?

File: lnrpc/invoicesrpc/addinvoice.go [hunk 7]
```diff
@@ -389,13 +392,17 @@ func AddInvoice(ctx context.Context, cfg *AddInvoiceConfig,
 		}
 
 		totalHopHints := len(invoice.RouteHints)
-		if invoice.Private {
+		if invoice.Private || invoice.HintsTarget > 0 {
 			totalHopHints = maxHopHints
 		}
 
 		hopHintsCfg := newSelectHopHintsCfg(cfg, totalHopHints)
+		nHintsTarget := 0
```
Comment (yyforyongyu) [inline lnrpc/invoicesrpc/addinvoice.go:400]: Looks like we can add the numHopHints field to `SelectHopHintsCfg` too, and no need to pass it to `PopulateHopHints`.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
The build failed🧐

File: cmd/lncli/cmd_invoice.go (hunk 4 above)
Comment (hotorb-code) [reply cmd/lncli/cmd_invoice.go:74]: `Hints` seems too similar to `RouteHints` but I also wanted the flag name to be short like the other flags.

**Comment by hotorb-code** [comment]
What is the correct way to compile the proto files? Is there a Docker image I should use?
I'm following the readmes and have Docker and Go installed. It seems to compile and format in Docker but even the source gives me formatting errors. Here's what I do:
`git clone https://github.com/lightningnetwork/lnd`
`cd lnd`
`make rpc-check`
Errors with "Protos not properly formatted or not compiled with v3.4.0".

File: lnrpc/lightning.proto (hunk 2 above)
Comment (hotorb-code) [reply lnrpc/lightning.proto]: Errors even with the unchanged source. What is the correct way to compile the proto files? Is there a Docker image I should use?

**Comment by guggero** [comment]
You need to run `make rpc` to compile them. The `make rpc-check` is for the CI to check that the protos were compiled.

**Comment by yyforyongyu** [comment]
Is this still under development? cc @hotorb-code

**Comment by hotorb-code** [comment]
I'm working on this again and made some changes.
Setting `hints` to 0 doesn't really make sense, right now it then adds the maximum amount of hints. Is that ok or should there be a different behavior?

**Comment by hotorb-code** [comment]
I'm working on this again and made some changes.

**Comment by yyforyongyu** [comment]
!lightninglabs-deploy mute

**Comment by ellemouton** [comment]
replaced by: https://github.com/lightningnetwork/lnd/pull/8819
